credential_provider/irsa_credential_provider.go

```diff
@@ -3,122 +3,117 @@ package credential_provider
 import (
 	"context"
 	"fmt"
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/credentials"
-	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
-	"github.com/aws/aws-sdk-go/aws/endpoints"
-	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials/stscreds"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
 	authv1 "k8s.io/api/authentication/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	k8sv1 "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/klog/v2"
 )
 
 const (
-	arnAnno      = "eks.amazonaws.com/role-arn"
-	docURL       = "https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html"
-	irsaAudience = "sts.amazonaws.com"
-	ProviderName = "secrets-store-csi-driver-provider-aws"
+	arnAnnotation = "eks.amazonaws.com/role-arn"
+	docURL        = "https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html"
+	irsaAudience  = "sts.amazonaws.com"
+	ProviderName  = "secrets-store-csi-driver-provider-aws"
 )
 
-// IRSACredentialProvider implements CredentialProvider using IAM Roles for Service Accounts
-type IRSACredentialProvider struct {
-	stsClient                 stsiface.STSAPI
-	k8sClient                 k8sv1.CoreV1Interface
-	region, nameSpace, svcAcc string
-	fetcher                   authTokenFetcher
-	ctx                       context.Context
+type irsaTokenFetcher struct {
+	namespace      string
+	serviceAccount string
+	k8sClient      k8sv1.CoreV1Interface
 }
 
-func NewIRSACredentialProvider(
-	stsClient stsiface.STSAPI,
-	region, nameSpace, svcAcc string,
-	k8sClient k8sv1.CoreV1Interface,
-	ctx context.Context,
-) CredentialProvider {
-	return &IRSACredentialProvider{
-		stsClient: stsClient,
-		k8sClient: k8sClient,
-		region:    region,
-		nameSpace: nameSpace,
-		svcAcc:    svcAcc,
-		fetcher:   newIRSATokenFetcher(nameSpace, svcAcc, k8sClient),
-		ctx:       ctx,
+func newIRSATokenFetcher(nameSpace, serviceAccount string, k8sClient k8sv1.CoreV1Interface) stscreds.IdentityTokenRetriever {
+	return &irsaTokenFetcher{
+		namespace:      nameSpace,
+		serviceAccount: serviceAccount,
+		k8sClient:      k8sClient,
 	}
 }
 
-func (p *IRSACredentialProvider) GetAWSConfig() (*aws.Config, error) {
-	roleArn, err := p.getRoleARN()
+// Private helper to fetch a JWT token for a given namespace and service account.
+//
+// See also: https://pkg.go.dev/k8s.io/client-go/kubernetes/typed/core/v1
+func (p *irsaTokenFetcher) GetIdentityToken() ([]byte, error) {
```

Comment:
This makes the token fetcher satisfy [the `stscreds.IdentityTokenRetriever` interface](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials@v1.17.59/stscreds#IdentityTokenRetriever), so we can pass that fetcher to [`stscreds.NewWebIdentityRoleProvider()`](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials@v1.17.59/stscreds#NewWebIdentityRoleProvider) later on during `aws.Config` construction